File: docs/api/examples/javascript/course-generation.js
/**
 * Viralify API - Course Generation Example (JavaScript/Node.js)
 *
 * This example demonstrates how to generate a complete video course
 * using the Viralify API.
 *
 * Requirements:
 *   npm install node-fetch dotenv
 *
 * Usage:
 *   export VIRALIFY_API_KEY="your_api_key"
 *   node course-generation.js
 */

import fetch from 'node-fetch';
import 'dotenv/config';

const API_KEY = process.env.VIRALIFY_API_KEY;
const BASE_URL = process.env.VIRALIFY_BASE_URL || 'https://api.viralify.io';

/**
 * Viralify API Client
 */
class ViralifyClient {
  constructor(apiKey, baseUrl = 'https://api.viralify.io') {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl;
  }

  /**
   * Make an API request
   */
  async request(method, endpoint, body = null) {
    const options = {
      method,
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
      },
    };

    if (body) {
      options.body = JSON.stringify(body);
    }

    const response = await fetch(`${this.baseUrl}${endpoint}`, options);

    if (!response.ok) {
      const error = await response.json();
      throw new Error(`API Error: ${error.message || response.statusText}`);
    }

    return response.json();
  }

  /**
   * Preview course outline
   */
  async previewOutline(config) {
    return this.request('POST', '/api/v1/courses/preview-outline', {
      topic: config.topic,
      difficulty_start: config.difficultyStart || 'beginner',
      difficulty_end: config.difficultyEnd || 'intermediate',
      structure: {
        number_of_sections: config.numSections || 4,
        lectures_per_section: config.lecturesPerSection || 3,
      },
      context: {
        category: config.category || 'tech',
      },
    });
  }

  /**
   * Start course generation
   */
  async generateCourse(config) {
    const result = await this.request('POST', '/api/v1/courses/generate', {
      topic: config.topic,
      difficulty_start: config.difficultyStart || 'beginner',
      difficulty_end: config.difficultyEnd || 'intermediate',
      structure: {
        number_of_sections: config.numSections || 4,
        lectures_per_section: config.lecturesPerSection || 3,
      },
      context: {
        category: config.category || 'tech',
      },
      language: config.language || 'en',
      quiz_config: {
        enabled: config.quizEnabled !== false,
        frequency: config.quizFrequency || 'per_section',
        questions_per_quiz: 5,
        passing_score: 70,
      },
      title_style: config.titleStyle || 'engaging',
      document_ids: config.documentIds || [],
    });

    return result.job_id;
  }

  /**
   * Get job status
   */
  async getJobStatus(jobId) {
    return this.request('GET', `/api/v1/courses/jobs/${jobId}`);
  }

  /**
   * Wait for job completion
   */
  async waitForCompletion(jobId, options = {}) {
    const pollInterval = options.pollInterval || 15000;
    const timeout = options.timeout || 3600000;
    const onProgress = options.onProgress || (() => {});

    const startTime = Date.now();

    while (true) {
      if (Date.now() - startTime > timeout) {
        throw new Error(`Job ${jobId} timed out`);
      }

      const status = await this.getJobStatus(jobId);
      onProgress(status);

      if (status.status === 'completed') {
        return status;
      }

      if (status.status === 'failed') {
        throw new Error(`Job failed: ${status.error || 'Unknown error'}`);
      }

      await this.sleep(pollInterval);
    }
  }

  /**
   * Upload document
   */
  async uploadDocument(filePath, userId, role = 'auto') {
    const fs = await import('fs');
    const path = await import('path');
    const FormData = (await import('form-data')).default;

    const form = new FormData();
    form.append('file', fs.createReadStream(filePath));
    form.append('user_id', userId);
    form.append('pedagogical_role', role);

    const response = await fetch(`${this.baseUrl}/api/v1/documents/upload`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: form,
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(`Upload failed: ${error.message}`);
    }

    return response.json();
  }

  /**
   * Sleep helper
   */
  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

/**
 * Progress callback
 */
function onProgress(status) {
  const progress = status.progress || 0;
  const stage = status.status || 'unknown';
  console.log(`[${stage.toUpperCase()}] Progress: ${progress.toFixed(1)}%`);
}

/**
 * Main function
 */
async function main() {
  if (!API_KEY) {
    console.error('Error: VIRALIFY_API_KEY environment variable not set');
    process.exit(1);
  }

  const client = new ViralifyClient(API_KEY, BASE_URL);

  // Course configuration
  const config = {
    topic: 'Introduction to React Hooks',
    difficultyStart: 'beginner',
    difficultyEnd: 'intermediate',
    numSections: 4,
    lecturesPerSection: 3,
    category: 'tech',
    language: 'en',
    quizEnabled: true,
    quizFrequency: 'per_section',
    titleStyle: 'engaging',
  };

  console.log(`\n=== Generating Course: ${config.topic} ===\n`);

  try {
    // Step 1: Preview outline
    console.log('Step 1: Previewing outline...');
    const outline = await client.previewOutline(config);
    console.log(`  Title: ${outline.title}`);
    console.log(`  Sections: ${outline.section_count}`);
    console.log(`  Total Lectures: ${outline.total_lectures}`);
    console.log();

    // Step 2: Start generation
    console.log('Step 2: Starting course generation...');
    const jobId = await client.generateCourse(config);
    console.log(`  Job ID: ${jobId}`);
    console.log();

    // Step 3: Wait for completion
    console.log('Step 3: Waiting for completion...');
    const result = await client.waitForCompletion(jobId, {
      pollInterval: 15000,
      timeout: 3600000,
      onProgress,
    });

    console.log('\n=== Course Generation Complete! ===');
    console.log(`Videos: ${result.output_urls.videos.length} files`);
    console.log(`ZIP: ${result.output_urls.zip}`);

    console.log('\nGenerated Videos:');
    result.output_urls.videos.forEach((url, i) => {
      console.log(`  ${i + 1}. ${url}`);
    });

  } catch (error) {
    console.error(`\nError: ${error.message}`);
    process.exit(1);
  }
}

main();

File: docs/api/examples/python/course_generation.py
"""
Viralify API - Course Generation Example (Python)

This example demonstrates how to generate a complete video course
using the Viralify API.

Requirements:
    pip install requests python-dotenv

Usage:
    export VIRALIFY_API_KEY="your_api_key"
    python course_generation.py
"""

import os
import time
import requests
from typing import Optional, Dict, Any
from dataclasses import dataclass


# Configuration
API_KEY = os.getenv("VIRALIFY_API_KEY")
BASE_URL = os.getenv("VIRALIFY_BASE_URL", "https://api.viralify.io")


@dataclass
class CourseConfig:
    """Configuration for course generation."""
    topic: str
    difficulty_start: str = "beginner"
    difficulty_end: str = "intermediate"
    num_sections: int = 4
    lectures_per_section: int = 3
    category: str = "tech"
    language: str = "en"
    quiz_enabled: bool = True
    quiz_frequency: str = "per_section"
    title_style: str = "engaging"


class ViralifyClient:
    """Simple Viralify API client."""

    def __init__(self, api_key: str, base_url: str = "https://api.viralify.io"):
        self.api_key = api_key
        self.base_url = base_url
        self.headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }

    def _request(self, method: str, endpoint: str, **kwargs) -> Dict[str, Any]:
        """Make an API request."""
        response = requests.request(
            method,
            f"{self.base_url}{endpoint}",
            headers=self.headers,
            timeout=60,
            **kwargs
        )
        response.raise_for_status()
        return response.json()

    def preview_outline(self, config: CourseConfig) -> Dict[str, Any]:
        """Preview course outline before generation."""
        return self._request("POST", "/api/v1/courses/preview-outline", json={
            "topic": config.topic,
            "difficulty_start": config.difficulty_start,
            "difficulty_end": config.difficulty_end,
            "structure": {
                "number_of_sections": config.num_sections,
                "lectures_per_section": config.lectures_per_section
            },
            "context": {
                "category": config.category
            }
        })

    def generate_course(self, config: CourseConfig, document_ids: Optional[list] = None) -> str:
        """Start course generation and return job ID."""
        payload = {
            "topic": config.topic,
            "difficulty_start": config.difficulty_start,
            "difficulty_end": config.difficulty_end,
            "structure": {
                "number_of_sections": config.num_sections,
                "lectures_per_section": config.lectures_per_section
            },
            "context": {
                "category": config.category
            },
            "language": config.language,
            "quiz_config": {
                "enabled": config.quiz_enabled,
                "frequency": config.quiz_frequency,
                "questions_per_quiz": 5,
                "passing_score": 70
            },
            "title_style": config.title_style
        }

        if document_ids:
            payload["document_ids"] = document_ids

        result = self._request("POST", "/api/v1/courses/generate", json=payload)
        return result["job_id"]

    def get_job_status(self, job_id: str) -> Dict[str, Any]:
        """Get course generation job status."""
        return self._request("GET", f"/api/v1/courses/jobs/{job_id}")

    def wait_for_completion(
        self,
        job_id: str,
        poll_interval: int = 15,
        timeout: int = 3600,
        callback=None
    ) -> Dict[str, Any]:
        """Wait for job completion with polling."""
        start_time = time.time()

        while True:
            if time.time() - start_time > timeout:
                raise TimeoutError(f"Job {job_id} timed out after {timeout}s")

            status = self.get_job_status(job_id)

            if callback:
                callback(status)

            if status["status"] == "completed":
                return status
            elif status["status"] == "failed":
                raise Exception(f"Job failed: {status.get('error', 'Unknown error')}")

            time.sleep(poll_interval)

    def upload_document(self, file_path: str, user_id: str, role: str = "auto") -> str:
        """Upload a document for RAG."""
        with open(file_path, "rb") as f:
            response = requests.post(
                f"{self.base_url}/api/v1/documents/upload",
                headers={"Authorization": f"Bearer {self.api_key}"},
                files={"file": f},
                data={"user_id": user_id, "pedagogical_role": role}
            )
            response.raise_for_status()
            return response.json()["id"]


def progress_callback(status: Dict[str, Any]):
    """Print progress updates."""
    progress = status.get("progress", 0)
    stage = status.get("status", "unknown")
    print(f"[{stage.upper()}] Progress: {progress:.1f}%")


def main():
    """Main example function."""
    if not API_KEY:
        print("Error: VIRALIFY_API_KEY environment variable not set")
        return

    client = ViralifyClient(API_KEY, BASE_URL)

    # Course configuration
    config = CourseConfig(
        topic="Introduction to Docker and Containerization",
        difficulty_start="beginner",
        difficulty_end="intermediate",
        num_sections=4,
        lectures_per_section=3,
        category="tech",
        language="en",
        quiz_enabled=True,
        quiz_frequency="per_section",
        title_style="engaging"
    )

    print(f"=== Generating Course: {config.topic} ===\n")

    # Step 1: Preview outline
    print("Step 1: Previewing outline...")
    outline = client.preview_outline(config)
    print(f"  Title: {outline['title']}")
    print(f"  Sections: {outline['section_count']}")
    print(f"  Total Lectures: {outline['total_lectures']}")
    print(f"  Duration: ~{outline.get('estimated_duration_minutes', 'N/A')} minutes")
    print()

    # Step 2: Start generation
    print("Step 2: Starting course generation...")
    job_id = client.generate_course(config)
    print(f"  Job ID: {job_id}")
    print()

    # Step 3: Wait for completion
    print("Step 3: Waiting for completion...")
    try:
        result = client.wait_for_completion(
            job_id,
            poll_interval=15,
            timeout=3600,
            callback=progress_callback
        )

        print("\n=== Course Generation Complete! ===")
        print(f"Videos: {len(result['output_urls']['videos'])} files")
        print(f"ZIP: {result['output_urls']['zip']}")

        # Print video URLs
        print("\nGenerated Videos:")
        for i, url in enumerate(result['output_urls']['videos'], 1):
            print(f"  {i}. {url}")

    except TimeoutError as e:
        print(f"Error: {e}")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()
